scipy/sparse/linalg/isolve/iterative.py

```diff
@@ -147,292 +301,199 @@ def combine(fn):
-        if callback is not None and iter_ > olditer:
-            callback(x)
-        slice1 = slice(ndx1-1, ndx1-1+n)
-        slice2 = slice(ndx2-1, ndx2-1+n)
-        if (ijob == -1):
-            if callback is not None:
-                callback(x)
-            break
-        elif (ijob == 1):
-            work[slice2] *= sclr2
-            work[slice2] += sclr1*matvec(work[slice1])
-        elif (ijob == 2):
-            work[slice1] = psolve(work[slice2])
-        elif (ijob == 3):
-            work[slice2] *= sclr2
-            work[slice2] += sclr1*matvec(x)
-        elif (ijob == 4):
-            if ftflag:
-                info = -1
-                ftflag = False
-            resid, info = _stoptest(work[slice1], atol)
-            if info == 1 and iter_ > 1:
-                # recompute residual and recheck, to avoid
-                # accumulating rounding error
-                work[slice1] = b - matvec(x)
-                resid, info = _stoptest(work[slice1], atol)
-        ijob = 2
-
-    if info == -10:
-        # termination due to breakdown: check for convergence
-        resid, ok = _stoptest(b - matvec(x), atol)
-        if ok:
-            info = 0
-
-    if info > 0 and iter_ == maxiter and not (resid <= atol):
-        # info isn't set appropriately otherwise
-        info = iter_
-
-    return postprocess(x), info
-
-
+def cgne(A, b, x0=None, tol=1e-5, maxit=10000, dtype=None, sparams=None, M=None):
+	assert A.shape[0] == A.shape[1] & A.shape[0] == len(b)
+	cpuKSPSolve = time.time()
+	A, M, x, b, postprocess = make_system(A, M, x0, b)
+	r = b - A.matvec(x)
+	At = A.H
+	z = M.matvec(r)
+	p = At.matvec(z)
+	ztr_old = np.inner(z.conjugate(), r)
+	r0norm = np.linalg.norm(r)
+	if r0norm == 0.:
+		return (postprocess(x), 0)
+	for iter in range(maxit):
+		w = A.matvec(p)
+		alpha = ztr_old / np.inner(p.conjugate(), p)
+		x += alpha * p
+		r -= alpha * w
+		alpha = np.linalg.norm(r) / r0norm
+		if sparams == '-scipy_ksp_monitor':
+			print("{}: unpreconditioned residual = {}".format(iter+1, alpha))
+		if alpha < tol:
+			print("Linear solve converged due to reaching TOL iterations {}".format(iter+1))
+			print(" --- system solved with SciPy/CGNR (in {})".format(time.time()-cpuKSPSolve))
+			return (postprocess(x), 0)
+		elif iter == maxit - 1:
+			print("Linear solve converged due to reaching MAXIT iterations {}".format(iter+1))
+			return (postprocess(x), 1)
+		z = M.matvec(r)
+		ztr = np.inner(z.conjugate(), r)
+		beta = ztr / ztr_old
+		p = At.matvec(z) + beta * p
+		ztr_old = ztr
+
+
+#======================
+#   Restarted GMRES
+#======================
 @non_reentrant()
-def gmres(A, b, x0=None, tol=1e-5, restart=None, maxiter=None, M=None, callback=None,
-          restrt=None, atol=None, callback_type=None):
-    """
-    Use Generalized Minimal RESidual iteration to solve ``Ax = b``.
+def gmres(A, b, x0=None, restart=None, tol=1e-5, maxit=10000, dtype=None, sparams=None, M=None):
```

Comment:
Isn't `gmres()` a public SciPy function? I think we usually try to remain backwards compatible and avoid things like changing `maxiter` parameter name to `maxit` and removing an argument completely as was done for `atol` here.

Maybe a private function or wrapper is needed? I'm not a domain expert though.

Reply:
Yeah, I am trying to add new gmres() implementation to replace original implementation (CPU time have an improvement). It works well for Poisson equations and convection-diffusion equations in 2D on my server, but it reports error when I submit it to community. As you suggested, I should use `maxiter` rather than `maxit`, and atol also should be added. Thanks.